File: src/csharp/examples/cudnn/Network.cs
using System;
using System.Linq;
using Alea.CUDA;
using Alea.CUDA.CULib;

namespace Tutorial.Cs.examples.cudnn
{
    //[mnistNetwork]
    public class Network : DisposableObject
    {
        private readonly Worker _worker;
        private readonly CUDNN _cudnn;
        private readonly CUBLAS _cublas;
        private readonly CUDNNTensorDescriptor _srcTensorDesc;
        private readonly CUDNNTensorDescriptor _dstTensorDesc;
        private readonly CUDNNTensorDescriptor _biasTensorDesc;
        private readonly CUDNNFilterDescriptor _filterDesc;
        private readonly CUDNNConvolutionDescriptor _convDesc;
        private readonly CUDNNPoolingDescriptor _poolingDesc;

        const CUDNNInterop.cudnnDataType_t DataType = CUDNNInterop.cudnnDataType_t.CUDNN_DATA_FLOAT;
        const CUDNNInterop.cudnnTensorFormat_t TensorFormat = CUDNNInterop.cudnnTensorFormat_t.CUDNN_TENSOR_NCHW;

        public Network(Worker worker)
        {
            _worker = worker;
            _cudnn = new CUDNN(_worker);
            _cublas = new CUBLAS(_worker);
            _srcTensorDesc = new CUDNNTensorDescriptor();
            _dstTensorDesc = new CUDNNTensorDescriptor();
            _biasTensorDesc = new CUDNNTensorDescriptor();
            _filterDesc = new CUDNNFilterDescriptor();
            _convDesc = new CUDNNConvolutionDescriptor();
            _poolingDesc = new CUDNNPoolingDescriptor();
        }

        // @Aaron: we need implement IDisposable, we have lot of unmanaged resources
        public override void Dispose(bool disposing)
        {
            if (disposing)
            {
                _poolingDesc.Dispose();
                _convDesc.Dispose();
                _filterDesc.Dispose();
                _biasTensorDesc.Dispose();
                _dstTensorDesc.Dispose();
                _srcTensorDesc.Dispose();
                _cublas.Dispose();
                _cudnn.Dispose();
            }
            base.Dispose(disposing);
        }

        public class nchw_t
        {
            public int N { get; set; }
            public int C { get; set; }
            public int H { get; set; }
            public int W { get; set; }
        }

        public void Resize(ref DeviceMemory<float> buffer, int length)
        {
            if (buffer.Length >= length) return;
            buffer.Dispose();
            buffer = _worker.Malloc<float>(length);
        }

        public void AddBias(CUDNNTensorDescriptor dstTensorDesc, Layer layer, int c, DeviceMemory<float> data)
        {
            _biasTensorDesc.Set4D(TensorFormat, DataType, 1, c, 1, 1);
            const float alpha = 1.0f;
            const float beta = 1.0f;
            _cudnn.AddTensor(CUDNNInterop.cudnnAddMode_t.CUDNN_ADD_SAME_C, alpha, _biasTensorDesc, layer.BiasD.Ptr, beta, dstTensorDesc, data.Ptr);
        }

        public void FullyConnectedForward(Layer ip, nchw_t nchw, DeviceMemory<float> srcData, ref DeviceMemory<float> dstData)
        {

            if (nchw.N != 1) throw new Exception("Not Implemented");
            //Console.WriteLine("fcf n,c,h,w ===> {0}, {1}, {2}, {3}", nchw.N, nchw.C, nchw.H, nchw.W);
            var dimX = nchw.C * nchw.H * nchw.W;
            var dimY = ip.Outputs;
            Resize(ref dstData, dimY);

            const float alpha = 1.0f;
            const float beta = 1.0f;

            // this cuMemcpyDtoD is raw CUDA API, should be guard with worker.eval
            var output = dstData;
            _worker.EvalAction(() => CUDAInterop.cuMemcpyDtoD(output.Ptr.Handle, ip.BiasD.Handle, (IntPtr)(dimY * sizeof(float))));

            // this doesn't need worker.eval, cause cudnn is a thin wrapper, it has worke.eval 
            _cublas.Sgemv(CUBLASInterop.cublasOperation_t.CUBLAS_OP_T, dimX, dimY, alpha, ip.DataD.Ptr, dimX,
                srcData.Ptr, 1, beta, dstData.Ptr, 1);

            nchw.H = 1;
            nchw.W = 1;
            nchw.C = dimY;
        }

        public void ConvoluteForward(Layer conv, nchw_t nchw, DeviceMemory<float> srcData, ref DeviceMemory<float> dstData)
        {
            //Console.WriteLine("n,c,h,w ======> {0}, {1}, {2}, {3}", nchw.N, nchw.C, nchw.H, nchw.W);
            _srcTensorDesc.Set4D(TensorFormat, DataType, nchw.N, nchw.C, nchw.H, nchw.W);
            _filterDesc.Set4D(DataType, conv.Outputs, conv.Inputs, conv.KernelDim, conv.KernelDim);
            _convDesc.Set2D(0, 0, 1, 1, 1, 1, CUDNNInterop.cudnnConvolutionMode_t.CUDNN_CROSS_CORRELATION);
            // find dimension of convoltion output
            // outputDim = 1 + (inputDim + 2*pad - filterDim) / convolutionStride
            int n, c, h, w;
            _convDesc.Get2DForwardOutputDim(_srcTensorDesc, _filterDesc, out n, out c, out h, out w);
            nchw.N = n;
            nchw.C = c;
            nchw.H = h;
            nchw.W = w;
            _dstTensorDesc.Set4D(TensorFormat, DataType, nchw.N, nchw.C, nchw.H, nchw.W);
            var algo = _cudnn.GetConvolutionForwardAlgorithm(_srcTensorDesc, _filterDesc, _convDesc, _dstTensorDesc,
                CUDNNInterop.cudnnConvolutionFwdPreference_t.CUDNN_CONVOLUTION_FWD_PREFER_FASTEST, (IntPtr)0);

            //Console.WriteLine("n,c,h,w ======> {0}, {1}, {2}, {3}", nchw.N, nchw.C, nchw.H, nchw.W);
            Resize(ref dstData, nchw.N * nchw.C * nchw.H * nchw.W);
            var sizeInBytes = _cudnn.GetConvolutionForwardWorkspaceSize(_srcTensorDesc, _filterDesc, _convDesc, _dstTensorDesc, algo);

            // I changed the malloc, so it can support malloc(0), where malloc(0).Ptr = 0n
            using (var workSpace = _worker.Malloc<byte>(sizeInBytes.ToInt32()))
            {
                const float alpha = 1.0f;
                const float beta = 0.0f;
                _cudnn.ConvolutionForward(alpha, _srcTensorDesc, srcData.Ptr, _filterDesc, conv.DataD.Ptr, _convDesc, algo, workSpace.Ptr, sizeInBytes, beta, _dstTensorDesc, dstData.Ptr);
                AddBias(_dstTensorDesc, conv, c, dstData);
            }
        }

        public void PoolForward(nchw_t nchw, DeviceMemory<float> srcData, ref DeviceMemory<float> dstData)
        {
            _poolingDesc.Set2D(CUDNNInterop.cudnnPoolingMode_t.CUDNN_POOLING_MAX, 2, 2, 0, 0, 2, 2);
            _srcTensorDesc.Set4D(TensorFormat, DataType, nchw.N, nchw.C, nchw.H, nchw.W);
            nchw.H /= 2;
            nchw.W /= 2;
            _dstTensorDesc.Set4D(TensorFormat, DataType, nchw.N, nchw.C, nchw.H, nchw.W);

            Resize(ref dstData, nchw.N * nchw.C * nchw.H * nchw.W);
            const float alpha = 1.0f;
            const float beta = 0.0f;
            //Console.WriteLine("n,c,h,w ======> {0}, {1}, {2}, {3}", nchw.N, nchw.C, nchw.H, nchw.W);
            _cudnn.PoolingForward(_poolingDesc, alpha, _srcTensorDesc, srcData.Ptr, beta, _dstTensorDesc, dstData.Ptr);
        }

        public void SoftmaxForward(nchw_t nchw, DeviceMemory<float> srcData, ref DeviceMemory<float> dstData)
        {
            Resize(ref dstData, nchw.N * nchw.C * nchw.H * nchw.W);
            _srcTensorDesc.Set4D(TensorFormat, DataType, nchw.N, nchw.C, nchw.H, nchw.W);
            _dstTensorDesc.Set4D(TensorFormat, DataType, nchw.N, nchw.C, nchw.H, nchw.W);
            const float alpha = 1.0f;
            const float beta = 0.0f;
            _cudnn.SoftmaxForward(CUDNNInterop.cudnnSoftmaxAlgorithm_t.CUDNN_SOFTMAX_ACCURATE, CUDNNInterop.cudnnSoftmaxMode_t.CUDNN_SOFTMAX_MODE_CHANNEL, alpha, _srcTensorDesc, srcData.Ptr, beta, _dstTensorDesc, dstData.Ptr);
        }

        public void ActivationForward(nchw_t nchw, DeviceMemory<float> srcData, ref DeviceMemory<float> dstData)
        {
            Resize(ref dstData, nchw.N * nchw.C * nchw.H * nchw.W);
            _srcTensorDesc.Set4D(TensorFormat, DataType, nchw.N, nchw.C, nchw.H, nchw.W);
            _dstTensorDesc.Set4D(TensorFormat, DataType, nchw.N, nchw.C, nchw.H, nchw.W);
            const float alpha = 1.0f;
            const float beta = 0.0f;
            _cudnn.ActivationForward(CUDNNInterop.cudnnActivationMode_t.CUDNN_ACTIVATION_RELU, alpha, _srcTensorDesc, srcData.Ptr, beta, _dstTensorDesc, dstData.Ptr);
        }

        public int ClassifyExample(string fname, Layer conv1, Layer conv2, Layer ip1, Layer ip2)
        {
            var nchw = new nchw_t()
            {
                N = 1,
                C = 1,
                H = Data.ImageH,
                W = Data.ImageW
            };

            var imgDataH = new float[Data.ImageH * Data.ImageW];
            var oHostSrc = Data.LoadImage(fname).Select(x => (int)x).ToArray();
            for (var i = 0; i < Data.ImageH; i++)
            {
                for (var j = 0; j < Data.ImageW; j++)
                {
                    var idx = Data.ImageH * i + j;
                    imgDataH[idx] = oHostSrc[idx] / 255.0f;
                }
            }

            using (var srcData = _worker.Malloc(imgDataH))
            using (var dstData = _worker.Malloc<float>(0))
            {
                Console.WriteLine("Performing forward propigation...");
                var src = srcData;
                var dst = dstData;

                ConvoluteForward(conv1, nchw, src, ref dst);
                PoolForward(nchw, dst, ref src);

                ConvoluteForward(conv2, nchw, src, ref dst);
                PoolForward(nchw, dst, ref src);

                FullyConnectedForward(ip1, nchw, src, ref dst);
                ActivationForward(nchw, dst, ref src);

                FullyConnectedForward(ip2, nchw, src, ref dst);
                SoftmaxForward(nchw, dst, ref src);

                Console.WriteLine("Finished forward propigation.");
                var maxDigits = 10;
                var hsrc = src.Gather();
                var result = hsrc.Take(maxDigits).ToArray();
                var id = 0;
                for (var i = 1; i < maxDigits; i++)
                    if (result[id] < result[i])
                        id = i;
                Console.WriteLine("Classification Complete.\n");
                return id;
            }

        }

    }
    //[/mnistNetwork]
}

File: src/fsharp/examples/cudnn/Network.fs
(*** hide ***)
module Tutorial.Fs.examples.cudnn.Network
open System
open System.IO
open Alea.CUDA
open Alea.CUDA.CULib
open Alea.CUDA.CULib.CUBLASInterop
open Alea.CUDA.CULib.CUDNNInterop
open Layer

type nchw_t = 
    {mutable N:int; mutable C:int; mutable H:int; mutable W:int}
    member x.set n c h w = x.N <- n; x.C <- c; x.H <- h; x.W <- w
    static member create n c h w = {N = n; C = c; H = h; W = w}

(*** define:CudnnMnistNetwork ***)
type Network(worker:Worker) =
    // It is a good idea to implement Network as a disposable object because we are using
    // many unmanaged resources.
    inherit DisposableObject()

    let DataType = CUDNNInterop.cudnnDataType_t.CUDNN_DATA_FLOAT;
    let TensorFormat = CUDNNInterop.cudnnTensorFormat_t.CUDNN_TENSOR_NCHW;
    
    let cudnn = new CUDNN(worker)
    let cublas = new CUBLAS(worker)
    
    let srcTensorDesc = new CUDNNTensorDescriptor()
    let dstTensorDesc = new CUDNNTensorDescriptor()
    let biasTensorDesc = new CUDNNTensorDescriptor()
    let filterDesc = new CUDNNFilterDescriptor()
    let convDesc = new CUDNNConvolutionDescriptor()
    let poolingDesc = new CUDNNPoolingDescriptor()
    
    member net.Resize(buffer:DeviceMemory<float32> ref, length:int) =
        if   buffer.contents.Length >= length
        then ()
        else buffer.contents.Dispose(); buffer := worker.Malloc<float32>(length)

    member net.AddBias(dstTensorDesc:CUDNNTensorDescriptor, layer:Layer, c:int, data:DeviceMemory<float32> ref) =
        biasTensorDesc.Set4D(TensorFormat, DataType, 1, c, 1, 1)
        let alpha, beta = 1.f, 1.f
        cudnn.AddTensor(CUDNNInterop.cudnnAddMode_t.CUDNN_ADD_SAME_C, alpha, biasTensorDesc, layer.BiasD.Ptr, beta, dstTensorDesc, data.contents.Ptr)

(*** define:CudnnMnistFCF ***)
    member net.FullyConnectedForward(ip:Layer, nchw:nchw_t, srcData:DeviceMemory<float32>, dstData:DeviceMemory<float32> ref) =
        if nchw.N <> 1 then failwith "Not Implemented"
        let dimX = nchw.C * nchw.H * nchw.W
        let dimY = ip.Outputs
        net.Resize(dstData, dimY)

        let alpha, beta = 1.f, 1.f
        // This cuMemcpyDtoD is a raw CUDA API call so it should be guarded with worker.Eval
        worker.Eval <| fun _ -> CUDAInterop.cuMemcpyDtoD(dstData.contents.Ptr.Handle, ip.BiasD.Handle, IntPtr(dimY * sizeof<float32>)) |> ignore
        // This cublas call doesn't need worker.Eval because cublas is a thin wrapper for the raw API 
        // and it alreadyhas worke.eval 
        cublas.Sgemv(CUBLASInterop.cublasOperation_t.CUBLAS_OP_T, dimX, dimY, alpha, ip.DataD.Ptr, dimX, srcData.Ptr, 1, beta, dstData.contents.Ptr, 1)
        nchw.H <- 1; nchw.W <- 1; nchw.C <- dimY

(*** define:CudnnMnistCF ***)
    member net.ConvoluteForward(conv:Layer, nchw:nchw_t, srcData:DeviceMemory<float32>, dstData:DeviceMemory<float32> ref) =
        srcTensorDesc.Set4D(TensorFormat, DataType, nchw.N, nchw.C, nchw.H, nchw.W)
        filterDesc.Set4D(DataType, conv.Outputs, conv.Inputs, conv.KernelDim, conv.KernelDim)
        convDesc.Set2D(0, 0, 1, 1, 1, 1, CUDNNInterop.cudnnConvolutionMode_t.CUDNN_CROSS_CORRELATION)
        // find dimension of convoltion output
        // outputDim = 1 + (inputDim + 2*pad - filterDim) / convolutionStride
        let n,c,h,w = convDesc.Get2DForwardOutputDim(srcTensorDesc, filterDesc)
        nchw.set n c h w
        dstTensorDesc.Set4D(TensorFormat, DataType, nchw.N, nchw.C, nchw.H, nchw.W)
        let algo = cudnn.GetConvolutionForwardAlgorithm(srcTensorDesc, filterDesc, convDesc, dstTensorDesc, CUDNNInterop.cudnnConvolutionFwdPreference_t.CUDNN_CONVOLUTION_FWD_PREFER_FASTEST, IntPtr 0)

        net.Resize(dstData, nchw.N * nchw.C * nchw.H * nchw.W)
        let sizeInBytes = cudnn.GetConvolutionForwardWorkspaceSize(srcTensorDesc, filterDesc, convDesc, dstTensorDesc, algo)
        use workSpace = worker.Malloc<byte>(sizeInBytes.ToInt32())
        let alpha, beta = 1.f, 0.f
        cudnn.ConvolutionForward(alpha, srcTensorDesc, srcData.Ptr, filterDesc, conv.DataD.Ptr, convDesc, algo, workSpace.Ptr, sizeInBytes, beta, dstTensorDesc, dstData.contents.Ptr)
        net.AddBias(dstTensorDesc, conv, nchw.C, dstData)

(*** define:CudnnMnistPF ***)        
    member net.PoolForward(nchw:nchw_t, srcData:DeviceMemory<float32>, dstData:DeviceMemory<float32> ref) =
        poolingDesc.Set2D(CUDNNInterop.cudnnPoolingMode_t.CUDNN_POOLING_MAX, 2, 2, 0, 0, 2, 2)
        srcTensorDesc.Set4D(TensorFormat, DataType, nchw.N, nchw.C, nchw.H, nchw.W)
        nchw.H <- nchw.H / 2
        nchw.W <- nchw.W / 2
        dstTensorDesc.Set4D(TensorFormat, DataType, nchw.N, nchw.C, nchw.H, nchw.W)
        net.Resize(dstData, nchw.N * nchw.C * nchw.H * nchw.W)
        let alpha, beta = 1.f, 0.f
        cudnn.PoolingForward(poolingDesc, alpha, srcTensorDesc, srcData.Ptr, beta, dstTensorDesc, dstData.Value.Ptr)

(*** define:CudnnMnistSF ***)
    member net.SoftmaxForward(nchw:nchw_t, srcData:DeviceMemory<float32>, dstData:DeviceMemory<float32> ref) =
        net.Resize(dstData, nchw.N * nchw.C * nchw.H * nchw.W)
        srcTensorDesc.Set4D(TensorFormat, DataType, nchw.N, nchw.C, nchw.H, nchw.W)
        dstTensorDesc.Set4D(TensorFormat, DataType, nchw.N, nchw.C, nchw.H, nchw.W)
        let alpha, beta = 1.f, 0.f
        cudnn.SoftmaxForward(CUDNNInterop.cudnnSoftmaxAlgorithm_t.CUDNN_SOFTMAX_ACCURATE, CUDNNInterop.cudnnSoftmaxMode_t.CUDNN_SOFTMAX_MODE_CHANNEL, alpha, srcTensorDesc, srcData.Ptr, beta, dstTensorDesc, dstData.contents.Ptr)

(*** define:CudnnMnistAF ***)
    member net.ActivationForward(nchw:nchw_t, srcData:DeviceMemory<float32>, dstData:DeviceMemory<float32> ref) =
        net.Resize(dstData, nchw.N * nchw.C * nchw.H * nchw.W)
        srcTensorDesc.Set4D(TensorFormat, DataType, nchw.N, nchw.C, nchw.H, nchw.W)
        dstTensorDesc.Set4D(TensorFormat, DataType, nchw.N, nchw.C, nchw.H, nchw.W)
        let alpha, beta = 1.f, 0.f
        cudnn.ActivationForward(CUDNNInterop.cudnnActivationMode_t.CUDNN_ACTIVATION_RELU, alpha, srcTensorDesc, srcData.Ptr, beta, dstTensorDesc, dstData.contents.Ptr)

    
    override net.Dispose(disposing:bool) =
        if disposing then
            cudnn.Dispose()
            cublas.Dispose()
            srcTensorDesc.Dispose()
            dstTensorDesc.Dispose()
            biasTensorDesc.Dispose()
            filterDesc.Dispose()
            convDesc.Dispose()
            poolingDesc.Dispose()

(*** define:CudnnMnistClassify ***)
    member net.ClassifyExample fname conv1 conv2 ip1 ip2 =
        let nchw = nchw_t.create 1 1 ImageH ImageW
        let imgDataH = Array.zeroCreate<float32> (ImageH * ImageW)
        let oHostSrc = fname |> loadImage |> Array.map (float32)
        for i = 0 to ImageH - 1 do
            for j = 0 to ImageW - 1 do
                let idx = ImageH*i+j
                imgDataH.[idx] <- oHostSrc.[idx] / 255.0f

        use srcData = worker.Malloc(imgDataH)
        use dstData = worker.Malloc<float32>(0)
        
        let src = ref srcData
        let dst = ref dstData
        
        printfn "Performing forward propigation..."
        
        net.ConvoluteForward(conv1, nchw, !src, dst)
        net.PoolForward(nchw, !dst, src)
        
        net.ConvoluteForward(conv2, nchw, !src, dst)
        net.PoolForward(nchw, !dst, src)
        
        net.FullyConnectedForward(ip1, nchw, !src, dst)
        net.ActivationForward(nchw, !dst, src)

        net.FullyConnectedForward(ip2, nchw, !src, dst)
        net.SoftmaxForward(nchw, !dst, src)
        
        printfn "Finished forward propigation."
        
        let maxDigits = 10;
        let result = src.contents.Gather().[0..maxDigits-1]
        let mutable id = 0
        for i = 1 to maxDigits - 1 do if result.[id] < result.[i] then id <- i
        printfn "Classification Complete.\n"
        id

